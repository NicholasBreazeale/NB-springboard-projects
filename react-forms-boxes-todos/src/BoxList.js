import React, { useState } from "react";
import { v4 as uuidv4 } from "uuid";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";

function BoxList() {
  const [boxes, setBoxes] = useState([]);

  function addBox({ color, width, height }) {
    setBoxes(bx => [...bx, {id: uuidv4(), color, width, height}]);
  }

  function removeBox(id) {
    setBoxes(boxes.filter(b => b.id !== id));
  }

  return (
    <>
      <NewBoxForm addBox={addBox} />
      <div data-testid="box-list">
        {boxes.map(b => 
          <Box backgroundColor={b.color} width={`${b.width}px`} height={`${b.height}px`} removeBox={() => removeBox(b.id)} key={b.id} />
        )}
      </div>
    </>
  );
}

export default BoxList;
