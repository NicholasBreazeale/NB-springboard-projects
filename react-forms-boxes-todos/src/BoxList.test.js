import React from "react";
import { fireEvent, render } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", function() {
  render(<BoxList />);
});

it("matches snapshot", function() {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add and remove boxes", function() {
  const { queryByLabelText, queryByTestId, asFragment } = render(<BoxList />);

  // no boxes should be rendered yet
  const boxList = queryByTestId("box-list");
  expect(boxList.children.length).toBe(0);

  const colorInput = queryByLabelText("Color:");
  const widthInput = queryByLabelText("Width:");
  const heightInput = queryByLabelText("Height:");
  const submitButton = queryByTestId("submit-btn");

  // create a new box
  fireEvent.change(colorInput, { target: { value: "#ff00ff" } });
  fireEvent.change(widthInput, { target: { value: "50" } });
  fireEvent.change(heightInput, { target: { value: "50" } });
  fireEvent.click(submitButton);

  // 1 box and its delete button should be rendered
  expect(asFragment()).toMatchSnapshot();

  // delete box and its button
  fireEvent.click(boxList.children[1]);
  expect(asFragment()).toMatchSnapshot();
});
