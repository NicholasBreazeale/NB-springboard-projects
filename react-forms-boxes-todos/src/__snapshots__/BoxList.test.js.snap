// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add and remove boxes 1`] = `
<DocumentFragment>
  <form>
    <label
      for="color"
    >
      Color:
    </label>
    <input
      id="color"
      name="color"
      type="color"
      value="#000000"
    />
    <label
      for="width"
    >
      Width:
    </label>
    <input
      id="width"
      name="width"
      type="number"
      value=""
    />
    <label
      for="height"
    >
      Height:
    </label>
    <input
      id="height"
      name="height"
      type="number"
      value=""
    />
    <button
      data-testid="submit-btn"
    >
      Add
    </button>
  </form>
  <div
    data-testid="box-list"
  >
    <div
      class="Box"
      style="background-color: rgb(255, 0, 255); width: 50px; height: 50px;"
    />
    <button>
      X
    </button>
  </div>
</DocumentFragment>
`;

exports[`can add and remove boxes 2`] = `
<DocumentFragment>
  <form>
    <label
      for="color"
    >
      Color:
    </label>
    <input
      id="color"
      name="color"
      type="color"
      value="#000000"
    />
    <label
      for="width"
    >
      Width:
    </label>
    <input
      id="width"
      name="width"
      type="number"
      value=""
    />
    <label
      for="height"
    >
      Height:
    </label>
    <input
      id="height"
      name="height"
      type="number"
      value=""
    />
    <button
      data-testid="submit-btn"
    >
      Add
    </button>
  </form>
  <div
    data-testid="box-list"
  />
</DocumentFragment>
`;

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="color"
    >
      Color:
    </label>
    <input
      id="color"
      name="color"
      type="color"
      value="#000000"
    />
    <label
      for="width"
    >
      Width:
    </label>
    <input
      id="width"
      name="width"
      type="number"
      value=""
    />
    <label
      for="height"
    >
      Height:
    </label>
    <input
      id="height"
      name="height"
      type="number"
      value=""
    />
    <button
      data-testid="submit-btn"
    >
      Add
    </button>
  </form>
  <div
    data-testid="box-list"
  />
</DocumentFragment>
`;
